win32ss/gdi/ntgdi/freetype.c

```diff
@@ -945,6 +1167,214 @@ SubstituteFontRecurse(LOGFONTW* pLogFont)
     return TRUE;    /* success */
 }
 
+static inline VOID
+FontLink_Chain_Init(
+    _Out_ PFONTLINK_CHAIN pChain,
+    _Inout_ PTEXTOBJ pTextObj,
+    _In_ FT_Face face,
+    _In_ PDC dc)
+{
+    RtlZeroMemory(pChain, sizeof(*pChain));
+    pChain->pBaseTextObj = pTextObj;
+    pChain->pDefFace = face;
+    pChain->dc = dc;
+}
+
+static NTSTATUS
+FontLink_Chain_LoadReg(
+    _Inout_ PFONTLINK_CHAIN pChain,
+    _Inout_ LPLOGFONTW pLF)
+{
+    NTSTATUS Status;
+    HKEY hKey;
+    DWORD cbData;
+    static const WCHAR s_szzDefFontLink[] =
+        L"tahoma.ttf,Tahoma\0"
+        L"msgothic.ttc,MS UI Gothic\0"
+        L"mingliu.ttc,PMingLiU\0"
+        L"simsun.ttc,SimSun\0"
+        L"gulim.ttc,Gulim\0"
+        L"\0";
+    static const WCHAR s_szzDefFixedPitchFontLink[] =
+        L"cour.ttf,Courier New\0"
+        L"msgothic.ttc,MS Gothic\0"
+        L"mingliu.ttc,MingLiU\0"
+        L"simsun.ttc,NSimSun\0"
+        L"gulim.ttc,GulimChe\0"
+        L"\0";
+    WCHAR szzFontLink[512]; // 512 is enough size
```

Comment:
```suggestion

    NTSTATUS Status;
    HKEY hKey;
    DWORD cbData;
    WCHAR szzFontLink[512]; // 512 is enough size
```

Reply:
> // 512 is enough size

Same comment as before; how do you know it's "enough size" ?
You should be able to RegQueryValue to retrieve the actual size of the data, validate it, allocate a correctly-sized buffer and call RegQueryValue if needed.

Reply:
Yeah but here it's a multi-string, so this means, that the user would have many "font1,font2\0" strings, possibly with already long font face names.

Reply:
Fixed.